compiler/src/test/java/com/squareup/anvil/compiler/ClassReferenceTest.kt

```diff
@@ -0,0 +1,291 @@
+    compile(
+      """
+      package com.squareup.test
+
+      class SomeClass1 {
+        fun test(string: String) = Unit
+      }
+
+      interface GenericInterface1<T> {
+        fun hello(): T
+      }
+
+      interface GenericInterface2<S> : GenericInterface1<S>
+
+      class SomeClass2 : GenericInterface1<String> {
+        override fun hello(): String = "hello"
+      }
+
+      class SomeClass3 : GenericInterface2<String> {
+        override fun hello(): String = "hello"
+      }
+      """,
+      allWarningsAsErrors = false,
+      codeGenerators = listOf(
+        simpleCodeGenerator { psiRef ->
+          val descriptorRef = psiRef.toDescriptorReference()
+
+          when (psiRef.shortName) {
+            "SomeClass1" -> {
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "test" }
+
+              assertThat(psiFunction.returnTypeOrNull()).isNull()
+              assertThat(descriptorFunction.returnType().fqName).isEqualTo(Unit::class.fqName)
+
+              // That is null as well, because it's not a generic return type and we can't resolve
+              // the return type from a function body. The compiler will take care of this in the
+              // backend.
+              assertThat(psiFunction.resolveGenericReturnTypeOrNull(psiRef)).isNull()
+            }
+            "SomeClass2", "SomeClass3" -> {
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "hello" }
+
+              assertThat(psiFunction.returnType().fqName).isEqualTo(FqName("kotlin.String"))
+              assertThat(descriptorFunction.returnType().fqName).isEqualTo(FqName("kotlin.String"))
+            }
+            "GenericInterface1" -> {
+              val psiFunction = psiRef.functions.single()
+              val descriptorFunction = descriptorRef.functions.single { it.name == "hello" }
+
+              assertThat(psiFunction.returnTypeOrNull()).isNull()
+              assertThat(descriptorFunction.returnTypeOrNull()).isNull()
+
+              val implementingClass2 = FqName("com.squareup.test.SomeClass2")
+                .toClassReference(psiRef.module)
+                .toPsiReference()
+
+              assertThat(
+                psiFunction.resolveGenericReturnType(implementingClass2).fqName
+              ).isEqualTo(FqName("kotlin.String"))
+
+              assertThat(
+                descriptorFunction.resolveGenericReturnType(implementingClass2).fqName
+              ).isEqualTo(FqName("kotlin.String"))
+
+              // TODO: This is currently failing, but should work in theory. The problem is a mix
+              //  of ClassReferences in the super types being PSI and Descriptor implementations.
+              //  We need to implement resolving types for Descriptors as well. That it's only
+              //  supported for PSI is a smell.
```

Comment:
That's something I'd like to work on. But it was very hard to test or even to verify. That lead me down the path of finding this way to "unit test" `ClassReference`.

Reply:
Getting closer to having a solution. Having unit tests is so much easier.